frr-tools/status/controller/bgpsessionstate_controller.go

```diff
@@ -0,0 +1,281 @@
+// SPDX-License-Identifier:Apache-2.0
+
+package controller
+
+import (
+	"context"
+	"net/netip"
+	"reflect"
+	"strings"
+	"time"
+
+	frrk8sv1beta1 "github.com/metallb/frr-k8s/api/v1beta1"
+	"github.com/metallb/frr-k8s/internal/frr"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+)
+
+const (
+	nodeLabel       = "frrk8s.metallb.io/node"
+	peerLabel       = "frrk8s.metallb.io/peer"
+	vrfLabel        = "frrk8s.metallb.io/vrf"
+	noBFDConfigured = "N/A"
+)
+
+type BGPPeersFetcher func() (map[string][]*frr.Neighbor, error)
+
+// BGPSessionStateReconciler reconciles a BGPSessionState object.
+type BGPSessionStateReconciler struct {
+	client.Client
+	BGPPeersFetcher
+	Logger       log.Logger
+	NodeName     string
+	Namespace    string
+	DaemonPod    *corev1.Pod
+	ResyncPeriod time.Duration
+}
+
+type peersStatuses map[string]*frrk8sv1beta1.BGPSessionState
+
+func (ps peersStatuses) add(s frrk8sv1beta1.BGPSessionState) {
+	ps[peerFor(s)] = s.DeepCopy()
+}
+
+func (ps peersStatuses) hasPeerFor(s frrk8sv1beta1.BGPSessionState) bool {
+	_, ok := ps[peerFor(s)]
+	return ok
+}
+
+func (ps peersStatuses) statusFor(peer string) *frrk8sv1beta1.BGPSessionState {
+	return ps[labelFormatForNeighbor(peer)]
+}
+
+func (ps peersStatuses) remove(peer string) {
+	delete(ps, labelFormatForNeighbor(peer))
+}
+
+// statusMapFor returns a map representation of vrf -> peers statuses for the given resources +
+// a slice for the duplicate resources that did not go into the map.
+func statusMapFor(statuses []frrk8sv1beta1.BGPSessionState) (map[string]peersStatuses, []frrk8sv1beta1.BGPSessionState) {
```

Comment:
let's call this peersStatusPerVRF.
The extra return argument looks a bit odd, but I guess we can live with it.

Please move this below as it's an util function

Reply:
done